File: courses/data_analysis/deepdive/pubsub-prework-solution/python/action_publisher.py
# Copyright 2017 Google Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
###############################################################################
from action_utils import ActionUtils
from google.cloud import pubsub_v1


# A basic Pub/Sub publisher for purposes of demonstrating use of the API.
class ActionPublisher:
  # Creates a new publisher associated with the given project and topic.
  def __init__(self, project, topic):
    self._publisher = pubsub_v1.PublisherClient()
    self._topic_path = self._publisher.topic_path(project, topic)

  def callback(self, message_future):
    # When timeout is unspecified, the exception method waits indefinitely.
    if message_future.exception(timeout=30):
      print('Publishing message on {} threw an Exception {}.'.format(
          topic_name, message_future.exception()))

  # Publishes the action on the topic associated with this publisher.
  def publish(self, action):
    data = ActionUtils.encode_action_as_json(action)
    data = data.encode('utf-8')

    message_future = self._publisher.publish(self._topic_path, data=data)
    message_future.add_done_callback(self.callback)

File: courses/data_analysis/deepdive/pubsub-exercises/exercise0-prework/java/solution/src/main/java/com/google/cloud/sme/pubsub/ActionPublisher.java
// Copyright 2017 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////
package com.google.cloud.sme.pubsub;

import com.google.api.core.ApiFuture;
import com.google.api.core.ApiFutureCallback;
import com.google.api.core.ApiFutures;
import com.google.api.gax.rpc.ApiException;
import com.google.cloud.pubsub.v1.Publisher;
import com.google.cloud.sme.common.ActionUtils;
import com.google.cloud.sme.Entities;
import com.google.pubsub.v1.PubsubMessage;
import com.google.pubsub.v1.ProjectTopicName;

/** A basic Pub/Sub publisher for purposes of demonstrating use of the API. */
public class ActionPublisher {
  private Publisher publisher = null;

  public ActionPublisher(String project, String topic) {
    ProjectTopicName fullTopic = ProjectTopicName.of(project, topic);
    Publisher.Builder builder = Publisher.newBuilder(fullTopic);
    try {
      this.publisher = builder.build();
    } catch (Exception e) {
      System.out.println("Could not create publisher: " + e);
    }
  }

  public void publish(Entities.Action action) {
    PubsubMessage message =
        PubsubMessage.newBuilder()
            .setData(ActionUtils.encodeActionAsJson(action))
            .build();
    ApiFuture<String> response = publisher.publish(message);
    ApiFutures.addCallback(response, new ApiFutureCallback<String>() {

      @Override
      public void onFailure(Throwable throwable) {
        if (throwable instanceof ApiException) {
          ApiException apiException = ((ApiException) throwable);
          // details on the API exception
          System.out.println(apiException.getStatusCode().getCode());
          System.out.println(apiException.isRetryable());
        }
        System.out.println("Error publishing message : " + message);
      }

      @Override
      public void onSuccess(String messageId) {
      }
    });
  }
}
